File: solutions/_2542_maximum_subsequence_score/solution.js
const { MinPriorityQueue } = require("datastructures-js");

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number}
 */
var maxScore = function (A, B, k) {
  let total = 0;
  let res = 0;
  const pairs = Array.from({ length: A.length }, (_, i) => [A[i], B[i]]);
  pairs.sort((a, b) => b[1] - a[1]);

  const pq = new MinPriorityQueue();
  for (const [a, b] of pairs) {
    pq.enqueue(a);
    total += a;
    if (pq.size() === k) {
      res = Math.max(res, total * b);
      total -= pq.dequeue();
    }
  }

  return res;
};

module.exports = { maxScore };

/*
https://leetcode.com/problems/maximum-subsequence-score/
*/

File: solutions/_2542_maximum_subsequence_score/solution.java

package _2542_maximum_subsequence_score;

import java.util.*;

class Solution {
    public long maxScore(int[] A, int[] B, int k) {
        long total = 0;
        long res = 0;
        int n = A.length;
        int[][] pairs = new int[n][2];
        for (int i = 0; i < n; i++) {
            pairs[i] = new int[] { A[i], B[i] };
        }
        Arrays.sort(pairs, (a, b) -> b[1] - a[1]);
        Queue<Integer> pq = new PriorityQueue<Integer>(k);
        for (int[] pair : pairs) {
            int a = pair[0];
            int b = pair[1];
            pq.add(a);
            total += a;
            if (pq.size() == k) {
                res = Math.max(res, total * b);
                total -= pq.poll();
            }
        }
        return res;
    }
}

/*
 * https://leetcode.com/problems/maximum-subsequence-score/
 */